assets/approval.py

```diff
@@ -0,0 +1,344 @@
+            approval_lock = self.engine.query(Approval)\
+                .filter(
+                    Approval.timestamp >= datetime.fromtimestamp(Decimal(version.get('timestamp'))),
+                    pool=self.pool)\
+                .index('ts-index')\
+                .first(desc=True)
+
+        metadata = []
+        if approval_lock:
+            for key in approval_lock.keys_():
+                value = getattr(approval_lock, key)
+                if type(value) is datetime:
+                    value = str(Decimal(value.timestamp()))
+                if type(value) is bool:
+                    value = str(value)
+                metadata.append(
+                    {
+                        'name': key,
+                        'value': value
+                    }
+                )
+
+            name_path = os.path.join(target_dir, 'name')
+            with open(name_path, 'w') as name:
+                name.write(getattr(approval_lock, 'lockname'))
+
+            metadata_path = os.path.join(target_dir, 'metadata')
+            with open(metadata_path, 'w') as metadata_file:
+                json.dump(metadata, metadata_file)
+        else:
+            log.info("No lock have been found")
+            exit(0)
+
+        return {
+            'version': {"timestamp": "{timestamp}".format(
+                timestamp=Decimal(getattr(approval_lock, 'timestamp').timestamp()))},
+            'metadata': metadata,
+        }
+
+    def query_lock(self, lock_name):
+        """
+        This method is used to query the lock in the approval loop to check if there is a change on it
+        :param lock_name: The name of the lock to fetch
+        :return: the dynamodb item
+        """
+        return self.engine.query(Approval) \
+            .filter(
+                lockname=lock_name,
+                pool=self.pool) \
+            .index('ts-index') \
+            .first(desc=True)
+
+    def out_cmd(self, target_dir, source, params):
+        """
+        This method is responsible to acquire or release a lock. If the lock doesn't exist yet, then the method
+        create it automatically.
+        If a lock is already acquired, the method will wait indefinitely until being able to acquire it.
+        :param target_dir:
+        :param source: is the same value as passed to check.
+        :param params: is an arbitrary JSON object passed along verbatim from params on a put.
+        :return: a dict with the version fetched and the metadata of the lock
+        """
+        metadata = []
+
+        if 'lock_name' not in params:
+            log.error('You must set a lock_name on params')
+        if 'action' not in params:
+            log.error('You must set an action on params')
+
+        approval_lock = self.query_lock(params['lock_name'])
+        need_approval = params.get('need_approval', False)
+
+        if 'claim' in params['action']:
```

Comment:
WDYT about split actions in private functions like `_do_claim` and `_do_release` ?